torchtext/csrc/bert_tokenizer.cpp

```diff
@@ -0,0 +1,290 @@
+#include <torchtext/csrc/bert_tokenizer.h>
+#include <utf8proc.h>
+
+namespace torchtext {
+
+std::string BERTEncoder::kUnkToken = "[UNK]";
+
+int kMaxCharsPerWords = 100;
+
+static bool _is_whitespace(uint32_t c) {
+  if (c == '\t' || c == '\n' || c == '\r' || c == ' ') {
+    return true;
+  }
+  return (UTF8PROC_CATEGORY_ZS == utf8proc_category(c));
+}
+
+static bool _is_control(uint32_t c) {
+  if (c == '\t' || c == '\n' || c == '\r') {
+    return false;
+  }
+  utf8proc_category_t cat = utf8proc_category(c);
+
+  // unicodedata return 'Cn' whereas utf8proc return 'Lo' for following unicode
+  // point Explicitly checking for this unicode point to avoid above
+  // discrepency.
+  if (c == 3332)
+    return true;
+  // Fixed: HF referece: All categories starting with 'C'
+  return (
+      cat == UTF8PROC_CATEGORY_CC || cat == UTF8PROC_CATEGORY_CF ||
+      cat == UTF8PROC_CATEGORY_CN || cat == UTF8PROC_CATEGORY_CS ||
+      cat == UTF8PROC_CATEGORY_CO);
+}
+
+static bool _is_chinese_char(uint32_t cp) {
+  if ((cp >= 0x4E00 && cp <= 0x9FFF) || (cp >= 0x3400 && cp <= 0x4DBF) ||
+      (cp >= 0x20000 && cp <= 0x2A6DF) || (cp >= 0x2A700 && cp <= 0x2B73F) ||
+      (cp >= 0x2B740 && cp <= 0x2B81F) || (cp >= 0x2B820 && cp <= 0x2CEAF) ||
+      (cp >= 0xF900 && cp <= 0xFAFF) || (cp >= 0x2F800 && cp <= 0x2FA1F)) {
+    return true;
+  }
+  return false;
+}
+
+static bool _is_punct_char(uint32_t cp) {
+  if ((cp >= 33 && cp <= 47) || (cp >= 58 && cp <= 64) ||
+      (cp >= 91 && cp <= 96) || (cp >= 123 && cp <= 126)) {
+    return true;
+  }
+
+  if (cp == ' ') {
+    return false;
+  }
+
+  int cate = static_cast<int>(utf8proc_category(cp));
+  return (cate >= 12 && cate <= 18);
+}
+
+static UString _convert_to_unicode(const std::string& text) {
+  size_t i = 0;
+  UString ret;
+  while (i < text.size()) {
+    uint32_t codepoint;
+    utf8proc_ssize_t forward = utf8proc_iterate(
+        (utf8proc_uint8_t*)&text[i],
+        text.size() - i,
+        (utf8proc_int32_t*)&codepoint);
+    if (forward < 0)
+      return UString();
+    ret.append(1, codepoint);
+    i += forward;
+  }
+  return ret;
+}
+
+static std::string _convert_from_unicode(const UString& text) {
+  char dst[64];
+  std::string ret;
+  for (auto ch : text) {
+    utf8proc_ssize_t num = utf8proc_encode_char(ch, (utf8proc_uint8_t*)dst);
+    if (num <= 0)
+      return "";
+    ret += std::string(dst, dst + num);
+  }
+  return ret;
+}
+
+static void to_lower(UString& text) {
+  for (size_t i = 0; i < text.size(); i++) {
+    text[i] = utf8proc_tolower(text[i]);
+  }
+}
+
+BERTEncoder::BERTEncoder(
+    const std::string& vocab_file,
+    bool do_lower_case,
+    c10::optional<bool> strip_accents)
+    : vocab_{_load_vocab_from_file(vocab_file, 1, 1)} {
```

Comment:
hmm, sure we can. The reason for vocab_ to be in initializer list is because it must be explicitly initialized as it not have copy constructor. But I guess we can move other objects too in initialization list although performance wise it should hardly make difference, it might be a good practice :).